base/fastmath.jl

```diff
@@ -0,0 +1,326 @@
+
+const fast_op =
+    Dict(# basic arithmetic
+         :+ => :add_fast,
+         :- => :sub_fast,
+         :* => :mul_fast,
+         :/ => :div_fast,
+         :(==) => :eq_fast,
+         :!= => :ne_fast,
+         :< => :lt_fast,
+         :<= => :le_fast,
+         :abs => :abs_fast,
+         :abs2 => :abs2_fast,
+         :cmp => :cmp_fast,
+         :conj => :conj_fast,
+         :inv => :inv_fast,
+         :mod => :mod_fast,
+         :rem => :rem_fast,
+         :sign => :sign_fast,
+         :isfinite => :isfinite_fast,
+         :isinf => :isinf_fast,
+         :isnan => :isnan_fast,
+         :issubnormal => :issubnormal_fast,
+         # math functions
+         :^ => :pow_fast,
+         :acos => :acos_fast,
+         :acosh => :acosh_fast,
+         :angle => :angle_fast,
+         :asin => :asin_fast,
+         :asinh => :asinh_fast,
+         :atan => :atan_fast,
+         :atan2 => :atan2_fast,
+         :atanh => :atanh_fast,
+         :cbrt => :cbrt_fast,
+         :cis => :cos_fast,
+         :cos => :cos_fast,
+         :cosh => :cosh_fast,
+         :exp10 => :exp10_fast,
+         :exp2 => :exp2_fast,
+         :exp => :exp_fast,
+         :expm1 => :expm1_fast,
+         :hypot => :hypot_fast,
+         :lgamma => :lgamma_fast,
+         :log10 => :log10_fast,
+         :log1p => :log1p_fast,
+         :log2 => :log2_fast,
+         :log => :log_fast,
+         :max => :max_fast,
+         :min => :min_fast,
+         :minmax => :minmax_fast,
+         :sin => :sin_fast,
+         :sinh => :sinh_fast,
+         :sqrt => :sqrt_fast,
+         :tan => :tan_fast,
+         :tanh => :tanh_fast)
+
+const rewrite_op =
+    Dict(:+= => :+,
+         :-= => :-,
+         :*= => :*,
+         :/= => :/,
+         :^= => :^)
+
+function make_fastmath(expr::Expr)
+    op = get(rewrite_op, expr.head, :nothing)
+    if op != :nothing
+        var = expr.args[1]
+        rhs = expr.args[2]
+        if isa(var, Symbol)
+            # simple assignment
+            expr = :($var = $(op)($var, $rhs))
+        elseif isa(var, Expr) && var.head == :ref
+            # array reference
+            arr = var.args[1]
+            inds = tuple(var.args[2:end]...)
+            arrvar = gensym()
+            indvars = tuple([gensym() for i in inds]...)
+            expr = :($arrvar = $arr
+                     $indvars = $inds
+                     $(arrvar)[indvars] = $(op)($(arrvar)[$indvars], $rhs))
+        end
+    end
+    Expr(make_fastmath(expr.head), map(make_fastmath, expr.args)...)
+end
+function make_fastmath(symb::Symbol)
+    fast_symb = get(fast_op, symb, :nothing)
+    if fast_symb == :nothing
+        return symb
+    end
+    :(Base.FastMath.$(fast_symb))
+end
+make_fastmath(expr) = expr
+
+macro fastmath(expr)
+    make_fastmath(esc(expr))
+end
+
+
+
+# Basic arithmetic
+
+@fastmath for T in (Float32, Float64)
+    @eval begin
+        sub_fast(x::$T) = box($T,Base.neg_float_fast(unbox($T,x)))
```

Comment:
These could written as `func_fast{T<:Union(Float32,Float64)}(x::T) = .....` instead of generating duplicate methods for each type.